### ISSUE TITLE
Congrats on 1.0, please avoid binary breaking changes 😊

### ISSUE THREAD

**Issue by cmeeren** [Tarmil/FSharp.SystemTextJson#132]
Congratulations on the 1.0 release! 🥳

I hope that the library will now avoid binary breaking changes (within each major version). For example, one binary breaking change I have encountered many times throughout 0.x, is the changing of method signatures; in particular, the addition of optional parameters to the `JsonFSharpConverter` constructor. This is not a source breaking change (will not cause a compiler error), but if projects A and B refer to two different versions of FSharp.SystemTextJson, and I refer to both A and B, then I will get errors at runtime because it can't find the expected method. (AFAIK, .NET uses the entire signature of the method as an "identifier" to pick one at runtime.)

**Comment by Tarmil** [comment]
Yeah, this is unfortunate. Either we use optional arguments like we currently do, but when adding a new option, we break binary compatibility; or we use settable properties, but making an options type mutable feels icky. I guess an F# equivalent of C# `init` properties would be convenient in this case.

Or we could introduce a builder class. That can be both immutable and binary-compatible. Maybe something like

```fsharp
let options = JsonSerializerOptions()

JsonFSharpOptions.Builder()
    .UnionEncoding(JsonUnionEncoding.ThothLike)
    .IncludeRecordProperties()
    .ToConverter()
|> options.Converters.Add
```

Or even with a convenient extension method on `JsonSerializerOptions`:

```fsharp
let options =
    JsonSerializerOptions()
        .AddFSharp(fun options ->
            options.UnionEncoding(JsonUnionEncoding.ThothLike)
                   .IncludeRecordProperties())
```

**Comment by cmeeren** [comment]
In lieu of `init`, I (as a package/API consumer) would have no problem with mutable options.

Alternatively, I have myself both consumed and developed builders like you suggest. However, you don't necessarily need a separate builder. You can just do:

> ```fsharp
> let options = JsonSerializerOptions()
> 
> JsonFSharpOptions()
>     .UnionEncoding(JsonUnionEncoding.ThothLike)
>     .IncludeRecordProperties()
> |> options.Converters.Add
> ```

The constructor would create a default options object, and each method call would return a new instance with the specified options using e.g. `{ options with ... }` internally (assuming you are using a record type; I haven't checked). No need for a separte builder class, `.Builder()`, or `.ToConverter()`.

**Comment by Tarmil** [comment]
> assuming you are using a record type; I haven't checked

I am not, but that can be changed internally, and then yes I can make `JsonFSharpOptions` itself the builder. And in F# 6, if we add an implicit cast to `JsonFSharpConverter`, it can be passed directly to `options.Converters.Add`.

**Comment by cmeeren** [comment]
> And in F# 6, if we add an implicit cast to `JsonFSharpConverter`, it can be passed directly to `options.Converters.Add`.

This has always worked. F# has always allowed using a subtype in place of a parent type in methods. No need to add an implicit cast for this.

**Comment by Tarmil** [comment]
`JsonFSharpOptions` isn't a subtype of anything though. `JsonFSharpConverter` is a subtype of `JsonConverterFactory`, but because of that inheritance, I'd rather not make it its own builder: there might be any amount of initialization happening in the `JsonConverterFactory` constructor, and that would be repeated in a pure builder.

**Comment by Tarmil** [comment]
The fluent options builder is now merged into master. See #141

**Comment by Tarmil** [comment]
The fluent options builder is released in v1.1.
